File: src/domain/clock.ts
/**
 * Derived game clock.
 *
 * The match clock is NEVER a free-running counter. Elapsed seconds at any
 * instant are recomputed from immutable clock-control events (`start` / `stop`)
 * in chronological order. A refresh or reconnect at the 73rd minute re-derives
 * the same value from the database.
 *
 * Inputs are plain data — no Drizzle or Next.js imports. These functions are
 * the only domain logic that decides "what time is it on the match clock?".
 */

export type ClockControl = 'start' | 'stop' | 'none';

export interface ClockEvent {
  /** Whether this event opens, closes, or has no effect on the clock. */
  clockControl: ClockControl;
  /** Wall-clock time at which the event was recorded. */
  wallTime: Date;
}

/**
 * Walk the events in time order and reduce them to (running, openSince).
 * O(n) single pass; ignores extra `start` while already running and extra
 * `stop` while not running so accidental double-taps don't break the clock.
 */
function reduceClock(events: readonly ClockEvent[]): {
  running: boolean;
  openSince: Date | null;
  closedSeconds: number;
} {
  let closedSeconds = 0;
  let openSince: Date | null = null;

  for (const ev of events) {
    if (ev.clockControl === 'start') {
      if (openSince === null) openSince = ev.wallTime;
    } else if (ev.clockControl === 'stop') {
      if (openSince !== null) {
        closedSeconds += (ev.wallTime.getTime() - openSince.getTime()) / 1000;
        openSince = null;
      }
    }
  }

  return { running: openSince !== null, openSince, closedSeconds };
}

/**
 * Elapsed match-clock seconds at `now`, given all prior clock-control events.
 *
 * `now` is required (not defaulted) so callers must pass the same reference
 * time used elsewhere in their request — keeps tests deterministic and avoids
 * accidental drift between displayed and stored values.
 *
 * Returns a non-negative integer. Out-of-order events are tolerated: if a
 * stop precedes its start due to clock skew, the segment is clamped to zero.
 */
export function elapsedSeconds(events: readonly ClockEvent[], now: Date): number {
  const { openSince, closedSeconds } = reduceClock(events);
  const openSeconds = openSince === null ? 0 : (now.getTime() - openSince.getTime()) / 1000;
  const total = closedSeconds + Math.max(0, openSeconds);
  return Math.max(0, Math.floor(total));
}

/** Whether the clock is currently running. O(n) single pass. */
export function isClockRunning(events: readonly ClockEvent[]): boolean {
  return reduceClock(events).running;
}

File: services/match-engine/internal/application/clock.go
// Pure clock derivation. Mirrors src/domain/clock.ts in touchline — same
// algorithm, same edge cases (ignores duplicate starts while running, ignores
// stops while stopped, clamps fractional seconds). A direct port keeps the
// engine's view of "what time is it on the match clock?" in sync with the
// client's locally-ticking GameClock.
package application

import (
	"math"
	"time"

	"github.com/jedi-knights/touchline/services/match-engine/internal/domain"
)

// ElapsedSeconds returns the match-clock seconds at `now`, given the
// chronological clock-control event log.
//
// O(n) single pass. Out-of-order events are tolerated: if a stop precedes
// its start due to clock skew, the segment is clamped to zero.
func ElapsedSeconds(events []domain.ClockEvent, now time.Time) int {
	openSince, closed := reduceClock(events)
	if openSince == nil {
		return clampInt(closed)
	}
	open := now.Sub(*openSince).Seconds()
	if open < 0 {
		open = 0
	}
	return clampInt(closed + open)
}

// IsClockRunning reports whether the clock is currently running.
func IsClockRunning(events []domain.ClockEvent) bool {
	openSince, _ := reduceClock(events)
	return openSince != nil
}

// CountStarts returns the number of clock_control=start events in the log.
// Used by the state transition logic to determine the current period.
func CountStarts(events []domain.ClockEvent) int {
	n := 0
	for _, e := range events {
		if e.ClockControl == domain.ClockStart {
			n++
		}
	}
	return n
}

// CountStops returns the number of clock_control=stop events in the log.
func CountStops(events []domain.ClockEvent) int {
	n := 0
	for _, e := range events {
		if e.ClockControl == domain.ClockStop {
			n++
		}
	}
	return n
}

// reduceClock walks the events and returns (openSince, closedSeconds). It
// tolerates the same edge cases as touchline's clock.ts:
//
//   - duplicate start while already running → ignored
//   - stop while not running                → ignored
func reduceClock(events []domain.ClockEvent) (*time.Time, float64) {
	var openSince *time.Time
	closed := 0.0
	for _, e := range events {
		switch e.ClockControl {
		case domain.ClockStart:
			if openSince == nil {
				t := e.WallTime
				openSince = &t
			}
		case domain.ClockStop:
			if openSince != nil {
				closed += e.WallTime.Sub(*openSince).Seconds()
				openSince = nil
			}
		case domain.ClockNone:
			// no effect
		}
	}
	return openSince, closed
}

func clampInt(f float64) int {
	if f < 0 {
		return 0
	}
	return int(math.Floor(f))
}
